stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+                // `wanted_tenures`. Start searching `next_wanted_tenures`.
+                0
+            };
+
+            while k < next_wanted_tenures.len() {
+                let bit = u16::try_from(k).expect("FATAL: more sortitions than u16::MAX");
+                if !next_invbits.get(bit).unwrap_or(false) {
+                    k += 1;
+                    continue;
+                }
+
+                // k now points to the item in wanted_tenures with the tenure-send block ID for
+                // `ch`.
+                test_debug!("next end tenure is k={} {:?}", k, &next_wanted_tenures[k]);
+                break;
+            }
+            let Some(wt_end) = next_wanted_tenures.get(k) else {
+                break;
+            };
+
+            let mut tenure_start_end = TenureStartEnd::new(
+                wt.tenure_id_consensus_hash.clone(),
+                wt_start.winning_block_id.clone(),
+                wt_end.winning_block_id.clone(),
+                rc,
+                pox_constants
+                    .block_height_to_reward_cycle(first_burn_height, wt_start.burn_height)
+                    .expect("FATAL: tenure from before system start"),
+                wt.processed,
+            );
+            tenure_start_end.fetch_end_block = true;
+            test_debug!(
+                "i={}, k={}, n={}, len={}, next_len={}; {:?}",
+                i,
+                k,
+                n,
+                wanted_tenures.len(),
+                next_wanted_tenures.len(),
+                &tenure_start_end
+            );
+            tenure_block_ids.insert(wt.tenure_id_consensus_hash.clone(), tenure_start_end);
+        }
+        Some(tenure_block_ids)
+    }
+}
+
+/// A set of confirmed downloader state machines assigned to one or more neighbors.  The block
+/// downloader runs tenure-downloaders in parallel, since the downloader for the N+1'st tenure
+/// needs to feed data into the Nth tenure.  This struct is responsible for scheduling peer
+/// connections to downloader state machines, such that each peer is assigned to at most one
+/// downloader.  A peer is assigned a downloader for the duration of at most one RPC request, at
+/// which point, it will be re-assigned a (possibly different) downloader.  As such, each machine
+/// can make progress even if there is only one available peer (in which case, that peer will get
+/// scheduled across multiple machines to drive their progress in the right sequence such that
+/// tenures will be incrementally fetched and yielded by the p2p state machine to the relayer).
+pub struct NakamotoTenureDownloaderSet {
+    /// A list of instantiated downloaders that are in progress
+    pub(crate) downloaders: Vec<Option<NakamotoTenureDownloader>>,
+    /// An assignment of peers to downloader machines in the `downloaders` list.
+    pub(crate) peers: HashMap<NeighborAddress, usize>,
```

Comment:
A more naive choice seems like it would be:

```suggestion
    /// An assignment of peers to downloader machines
    pub(crate) downloaders: HashMap<NeighborAddress, NakamotoTenureDownloader>,
```

Or, if you wanted to differentiate store peers without active downloaders:

```rust
    pub(crate) downloaders: HashMap<NeighborAddress, Option<NakamotoTenureDownloader>>,
```

This removes any possibility of struct incoherence (i.e., a `peers` entry to the wrong downloader, or outside the bounds of `downloaders`, one of which would lead to a panic), and would also simplify the code in this struct  substantially (there's a lot of `self.downloaders.get(self.peers.get(naddr))` in the impl). Is there a reason a vec and pointer map is preferable?

Reply:
Oh, nevermind, I see why this is being done.

I think that there's a couple things that should be done as safeguards here:

1. Both these fields must be private. They are intrinsically tied to one another, and really shouldn't be modified except through these instance methods.
2. This is already the case-- `self.downloaders` should not ever have an element removed. But, the rustdocs on these fields need to make this very clear.

Reply:
They are `pub(crate)` because the test module needs to access them.  If you know of a better way to scope fields so that a unit test in a separate file can access them, I'm all ears.